lms/views/api/lti.py

```diff
@@ -114,8 +114,8 @@ def get_speedgrader_launch_url(self):
 
         if parsed_params.get("document_url"):
             params["url"] = parsed_params.get("document_url")
-
-        elif parsed_params.get("canvas_file_id"):
+        else:
+            assert parsed_params.get("canvas_file_id")
```

Comment:
I seem to remember having a conversation about using asserts in production code, and how it's not a thing we do.

I'm totally fine with it for sanity style this code should never get here style things. It seems to me that this could quite easily get in this state though if we get a `document_url` and no `canvas_file_id`.

What's the error handling for `AssertionError`? Will it result in a 400? Previously this wouldn't have failed.

Reply:
No, I think `assert`s are fine in production code as internal consistency checks and documentation, as long as they're used for unrecoverable errors / internal self-checks about conditions that should be impossible (and not e.g. for validation). Python's `-O` and `-OO` command line arguments disable asserts, but we don't use those. We didn't have a conversation in which I said we don't do them. This is a good post on it: https://dbader.org/blog/python-assert-tutorial

An uncaught `AssertionError` will be caught by the `Exception` exception view, so it'll be a 500 Internal Server Error, the user will get the HTML or JSON error response, and it'll be both logged to Papertrail and reported to Sentry.

Previously this code wouldn't have crashed but it would have generated an invalid SpeedGrader launch URL (one with neither a `url` or a `file_id` query param) and sent it to SpeedGrader, and SpeedGrader would then launch us with that invalid URL, the SpeedGrader launch-handling code isn't expecting such invalid launches so it would either misbehave or crash. Crashing early seems a lot less bad.

I'll definitely test SpeedGrader before deploying this though

Reply:
This is coming from Canvas anyway, so it's not like a user is typing this stuff in. If the previous state was a failure state, that doesn't worry me so much. I just wanted to check we weren't introducing one, as this code didn't previously make it obvious this wasn't an acceptable path.

The message is good. We can look for it in sentry, and fix it if it ever happens.